File: risk_core_cpp/src/validators.cpp
#include "risk_core/validators.hpp"
#include <algorithm>
#include <cctype>

namespace risk_core {

void validate_currency_code(const std::string& currency) {
    if (currency.empty() || currency.length() != 3) {
        throw ValidationError("Invalid currency code: " + currency + ". Must be 3 characters (ISO 4217)");
    }
    
    for (char c : currency) {
        if (!std::isalpha(c) || !std::isupper(c)) {
            throw ValidationError("Invalid currency code: " + currency + ". Must be uppercase letters");
        }
    }
}

void validate_asset(const Asset& asset) {
    if (asset.id().empty()) {
        throw ValidationError("Asset ID cannot be empty");
    }
    
    validate_currency_code(asset.currency());
    
    if (asset.value() < 0.0) {
        throw ValidationError("Asset value cannot be negative: " + std::to_string(asset.value()));
    }
}

void validate_portfolio(const Portfolio& portfolio) {
    if (portfolio.assets().empty()) {
        throw ValidationError("Portfolio must contain at least one asset");
    }
    
    std::vector<std::string> asset_ids;
    for (const auto& asset : portfolio.assets()) {
        validate_asset(asset);
        
        if (std::find(asset_ids.begin(), asset_ids.end(), asset.id()) != asset_ids.end()) {
            throw ValidationError("Duplicate asset ID: " + asset.id());
        }
        asset_ids.push_back(asset.id());
    }
}

void validate_scenario(const Scenario& scenario) {
    // Validate basis points are within reasonable range
    // Typically -10000 to +10000 bps (-100% to +100%)
    const double max_bps = 10000.0;
    const double min_bps = -10000.0;
    
    if (scenario.market_movement_bps() < min_bps || scenario.market_movement_bps() > max_bps) {
        throw ValidationError("Market movement BPS out of range: " + 
                             std::to_string(scenario.market_movement_bps()) +
                             ". Must be between " + std::to_string(min_bps) + 
                             " and " + std::to_string(max_bps));
    }
    
    if (scenario.volatility_change_bps() < min_bps || scenario.volatility_change_bps() > max_bps) {
        throw ValidationError("Volatility change BPS out of range: " + 
                             std::to_string(scenario.volatility_change_bps()) +
                             ". Must be between " + std::to_string(min_bps) + 
                             " and " + std::to_string(max_bps));
    }
    
    if (scenario.interest_rate_change_bps() < min_bps || scenario.interest_rate_change_bps() > max_bps) {
        throw ValidationError("Interest rate change BPS out of range: " + 
                             std::to_string(scenario.interest_rate_change_bps()) +
                             ". Must be between " + std::to_string(min_bps) + 
                             " and " + std::to_string(max_bps));
    }
    
    // Validate currency rate changes
    for (const auto& pair : scenario.currency_rate_changes()) {
        validate_currency_code(pair.first);
        if (pair.second < min_bps || pair.second > max_bps) {
            throw ValidationError("Currency rate change BPS for " + pair.first + 
                                 " out of range: " + std::to_string(pair.second) +
                                 ". Must be between " + std::to_string(min_bps) + 
                                 " and " + std::to_string(max_bps));
        }
    }
}

} // namespace risk_core

File: risk_core/validators.py
"""
Input validation functions for assets, portfolios, and scenarios.
"""

from decimal import Decimal
from typing import List, Dict
from .models import Asset, Portfolio, Scenario


class ValidationError(Exception):
    """Raised when validation fails."""
    pass


def validate_currency_code(currency: str) -> bool:
    """
    Validate currency code format (ISO 4217).
    
    Args:
        currency: Currency code to validate
        
    Returns:
        True if valid
        
    Raises:
        ValidationError: If currency code is invalid
    """
    if not currency or len(currency) != 3:
        raise ValidationError(f"Invalid currency code: {currency}. Must be 3 characters (ISO 4217)")
    
    if not currency.isalpha() or not currency.isupper():
        raise ValidationError(f"Invalid currency code: {currency}. Must be uppercase letters")
    
    return True


def validate_asset(asset: Asset) -> None:
    """
    Validate an asset object.
    
    Args:
        asset: Asset to validate
        
    Raises:
        ValidationError: If asset is invalid
    """
    if not asset.id or not asset.id.strip():
        raise ValidationError("Asset ID cannot be empty")
    
    validate_currency_code(asset.currency)
    
    if asset.value < Decimal('0'):
        raise ValidationError(f"Asset value cannot be negative: {asset.value}")


def validate_portfolio(portfolio: Portfolio) -> None:
    """
    Validate a portfolio object.
    
    Args:
        portfolio: Portfolio to validate
        
    Raises:
        ValidationError: If portfolio is invalid
    """
    if not portfolio.assets:
        raise ValidationError("Portfolio must contain at least one asset")
    
    asset_ids = set()
    for asset in portfolio.assets:
        validate_asset(asset)
        
        if asset.id in asset_ids:
            raise ValidationError(f"Duplicate asset ID: {asset.id}")
        asset_ids.add(asset.id)


def validate_scenario(scenario: Scenario) -> None:
    """
    Validate a scenario object.
    
    Args:
        scenario: Scenario to validate
        
    Raises:
        ValidationError: If scenario is invalid
    """
    # Validate basis points are within reasonable range
    # Typically -10000 to +10000 bps (-100% to +100%)
    max_bps = Decimal('10000')
    min_bps = Decimal('-10000')
    
    if scenario.market_movement_bps < min_bps or scenario.market_movement_bps > max_bps:
        raise ValidationError(
            f"Market movement BPS out of range: {scenario.market_movement_bps}. "
            f"Must be between {min_bps} and {max_bps}"
        )
    
    if scenario.volatility_change_bps < min_bps or scenario.volatility_change_bps > max_bps:
        raise ValidationError(
            f"Volatility change BPS out of range: {scenario.volatility_change_bps}. "
            f"Must be between {min_bps} and {max_bps}"
        )
    
    if scenario.interest_rate_change_bps < min_bps or scenario.interest_rate_change_bps > max_bps:
        raise ValidationError(
            f"Interest rate change BPS out of range: {scenario.interest_rate_change_bps}. "
            f"Must be between {min_bps} and {max_bps}"
        )
    
    # Validate currency rate changes
    for currency, rate_change in scenario.currency_rate_changes.items():
        validate_currency_code(currency)
        if rate_change < min_bps or rate_change > max_bps:
            raise ValidationError(
                f"Currency rate change BPS for {currency} out of range: {rate_change}. "
                f"Must be between {min_bps} and {max_bps}"
            )

